ci(docs-infra): move jobs to CircleCI

gkalpak opened the pull request:
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] ~~Docs have been added / updated (for bug fixes / features)~~


## PR Type
```
[x] CI related changes
```

## What is the current behavior?
Most `aio`-related CI jobs (such as test and deploy) run on Travis.


## What is the new behavior?
All `aio`-related CI jobs run on CircleCI.


## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

---
## TODOs:

✅ Deal with CircleCI flakes in `test_docs_examples` jobs 😞
✅ Ensure `payload-size` check works on CircleCI and uncomment in CircleCI config. (See also https://github.com/angular/angular/pull/26377#discussion_r225457091.)
    _As discussed, we will only use `$CI_COMMIT_RANGE` on non-PR builds, where we can infer it from CircleCI's `$CIRCLE_COMPARE_URL`._
✅ Ensure `aio/scripts/deploy-to-firebase.sh` works on CircleCI and move `aio` deployment from Travis to CircleCI.
✅ Remove `aio` jobs (and unused secrets) from Travis config.
✅ Uncomment temporarily commented out jobs in CircleCI config.
✅ Update once #26515 has been merged (e.g. https://github.com/angular/angular/pull/26515/commits/6cec8b54487c7c57bf0aa5445d06458c7fc7dbc8#diff-0610f203fb2c62d1b810313d19e10347R13).
✅ Provide `$ANGULAR_PAYLOAD_FIREBASE_TOKEN` to CircleCI (for non-PR builds). (Used in: `scripts/ci/payload-size.sh`)
    (Travis token was generated with `firebase login:ci` using the `payload@angular.io` account and password from valentine.)
✅ Provide `$FIREBASE_TOKEN` to CircleCI (for non-PR builds). (Used in: `aio/scripts/deploy-to-firebase.sh`)
    (Travis token was generated with `firebase login:ci` using the `aio-deploy@angular.io` account and password from valentine.)

##
Optinal TODOs:

✅ Use `--progress=false` for `yarn build[-local]` on CI.

##
Opened #26596 to explore possible further improvements (caching, sharing build artifacts, using parameterized jobs, etc.).

gkalpak committed:
test(docs-infra): improve logging output in `test-pwa-score[-localhost]`

gkalpak committed:
ci(docs-infra): add jobs to CircleCI config

petebacondarwin commented:
Fixes #17552

gkalpak commented:
It improves times a lot, but I don't think it fixes #17552 (see https://github.com/angular/angular/issues/17552#issuecomment-429329285).
Let's discuss there.

IgorMinar commented on .circleci/config.yml:
```diff
@@ -140,6 +140,106 @@ jobs:
+        # Run examples tests
+      - run: xvfb-run --auto-servernum yarn --cwd aio example-e2e --setup --local --shard=1=0/2
+
+  test_aio_examples_1:
```
can we rename this to "test_docs_examples_1" so that it's clear that this is not something specific the to the aio app but rather a generic doc app? thanks!

IgorMinar commented:
it's not quite clear to me what's left to do in this PR? @gkalpak could you please make a todo list?
one thing that I noticed is that we should remove these tests from the travis config....

gkalpak commented on .circleci/config.yml:
```diff
@@ -140,6 +140,106 @@ jobs:
+            xvfb-run --auto-servernum yarn --cwd aio test-pwa-score-localhost $AIO_MIN_PWA_SCORE
+        # Check the bundle sizes.
+        # (Run before unit and e2e tests, which destroy the `dist/` directory.)
+    #   - run: yarn --cwd aio payload-size
```
I temporarily commented it out, because this is currently failing on CircleCI due to missing env vars. The payload script is using Travis-specific vars (which is easy to replace with CircleCI ones, except for `TRAVIS_COMMIT_RANGE`, for which CircleCI does not provide a replacement and we haven't found a good alternative afaik - #23558 is a possible work around, but might be a little too heavy-handed (plus is subject to GitHub API's rate-limiting)).

```
$TRAVIS_BRANCH       --> $CIRCLE_BRANCH
$TRAVIS_COMMIT       --> $CIRCLE_SHA1
$TRAVIS_PULL_REQUEST --> $CIRCLE_PR_NUMBER (possibly different values when not a PR: `''` vs `'false'`)
$TRAVIS_COMMIT_RANGE --> ??
```

gkalpak commented on .circleci/config.yml:
```diff
@@ -266,36 +366,49 @@ workflows:
-      - test
-      - test_ivy_jit
-      - test_ivy_aot
+    #   - lint
```
These are just temporarily disabled, so that only the `aio` jobs that I care about run on CI (to save resources and speed things up).
I will uncomment once the PR is ready.

gkalpak commented:
This is marked as `WIP` (just sayin' 😛)

Anyway, here's a TODO list:

- [ ] Fix and re-enable `payload-size` check (see https://github.com/angular/angular/pull/26377#discussion_r225457091).
- [ ] Deal with CircleCI flakes in `test_docs_examples` jobs 😞
- [ ] Remove `aio` jobs from Travis config.
- [ ] Uncomment temporarily commented out jobs

My main blocker is how to handle the fact that CircleCI does not provide an alternative for `TRAVIS_COMMIT_RANGE` 😒

gkalpak committed:
test(animations): make e2e tests for animations docs examples less flaky

gkalpak committed:
test(elements): make e2e tests for elements docs examples less flaky

IgorMinar commented:
Didn't we have a solution for TRAVIS_COMMIT_RANGE on CircleCI already? maybe @petebacondarwin or @alexeagle would know...

gkalpak commented:
We have [this][1], which kind of emulates `$TRAVIS_COMMIT_RANGE` by guessing the base branch based on the version in `package.json`.
Maybe it is good enough for `payload-size` check.

[1]: https://github.com/angular/angular/blob/0f7d2ca7a81b9951863944e173ea77c783e423c0/tools/gulp-tasks/validate-commit-message.js#L15-L39

gkalpak committed:
build: use CI-provider independent variable names

gkalpak committed:
ci(docs-infra): move deployment to CircleCI

gkalpak committed:
ci: fix `payload-size.sh`

mary-poppins commented:
You can preview 39a3d14 at https://pr26377-39a3d14.ngbuilds.io/.

mary-poppins commented:
You can preview 5c4351a at https://pr26377-5c4351a.ngbuilds.io/.

mary-poppins commented:
You can preview d3d2566 at https://pr26377-d3d2566.ngbuilds.io/.

gkalpak committed:
ci(docs-infra): remove jobs from Travis config

mary-poppins commented:
You can preview 373f314 at https://pr26377-373f314.ngbuilds.io/.

mary-poppins commented:
You can preview 5f47be3 at https://pr26377-5f47be3.ngbuilds.io/.

gkalpak commented:
New issue:

✔️ `$TRAVIS_BRANCH` for PRs: The branch being targeted by the PR (e.g. `master` or `6.1.x`).
❌ `$CIRCLE_BRANCH` for PRs: The branch being built (e.g. `pull/12345`).

:unamused:

mary-poppins commented:
You can preview 4ee29aa at https://pr26377-4ee29aa.ngbuilds.io/.

mary-poppins commented:
You can preview ee41834 at https://pr26377-ee41834.ngbuilds.io/.

mary-poppins commented:
You can preview 30e7a3d at https://pr26377-30e7a3d.ngbuilds.io/.

mary-poppins commented:
You can preview 09028d6 at https://pr26377-09028d6.ngbuilds.io/.

mary-poppins commented:
You can preview a2c507d at https://pr26377-a2c507d.ngbuilds.io/.

mary-poppins commented:
You can preview 2306d1a at https://pr26377-2306d1a.ngbuilds.io/.

mary-poppins commented:
You can preview bd2e8e0 at https://pr26377-bd2e8e0.ngbuilds.io/.

petebacondarwin commented on .circleci/config.yml:
```diff
@@ -17,7 +17,20 @@ var_2: &cache_key v2-angular-{{ .Branch }}-{{ checksum "yarn.lock" }}-0.6.0
+      echo "export CI_COMMIT_RANGE=$(sed -r 's|^.*/([0-9a-f]+\.\.\.[0-9a-f]+)$|\1|i' <<< ${CIRCLE_COMPARE_URL:-})" | tee -a $BASH_ENV
+      echo "export CI_PULL_REQUEST=${CIRCLE_PR_NUMBER:-false}" | tee -a $BASH_ENV
+      echo "export CI_REPO_NAME=$CIRCLE_PROJECT_REPONAME" | tee -a $BASH_ENV
+      echo "export CI_REPO_OWNER=$CIRCLE_PROJECT_USERNAME" | tee -a $BASH_ENV
```
Is there a reason for putting these here rather than sourcing a shell script?

gkalpak replied:
Is there a reason for sourcing a shell script rather than putting these here? 😛

Kidding aside, here is my reasoning:
The values of most of these variables are CI-dependent. So, even if we put them in a source-able script, it would not really be usesable outside of CircleCI.
There are some variables that are independent of CI (e.g. `CI_AIO_MIN_PWA_SCORE`), and which we could put in a source-able, re-usable script, but they are so few that I thought it is better to just put them here.

It was kind of painful to switch from `TRAVIS_*` vars to `CIRCLE_*` vars, so I think everything will be better in the future if:
a. We use CI-provider-independent names for env vars.
b. Define all our "global" vars (ehm...I mean...constants) in one place (either a script or the top of `config.yml` work equally well for me :grin:).

This way, switching to or adding a new CI-provider is just a matter of defining those constants in a place that makes sense for that specific CI-provider.

Happy to take any alternative approach, as long as we don't scatter CI-provider-specific vars across our scripts :grin:

petebacondarwin replied:
OK, you convinced me :-)

IgorMinar replied:
In the past `env.sh` was meant to be the script for these variables. using echo and tee on every line is a smell imo. could we should source a single circleci-env.sh file instead?

IgorMinar replied:
I'm not convinced :) every line contains "echo" and "tee" which make it harder for anyone to understand what's going on. this is why I created env.sh file for travis. don't you think that having circleci-env.sh would make this more readable?

gkalpak replied:
I am not sure exactly what is your concern 😕

1. I was thinking about having a bash function to set env vars (similar to `env.sh`'s `setEnvVar`), which would take care of the repetitive `echo`-ing and `tee`-ing.

2. And we could move the code in a `.sh` file and source that, but it wouldn't change much other than moving the code out of `config.yml`.
(I.e. there would still be `echo`-ing and `tee`-ing (or something similar) and the `.sh` file will not be useable outside of CircleCI.)

##
If you just want to "hide away" this possibly confusing code block, then moving it to a `.sh` file is fine.
But other than that, I don't see any other benefit (and I would say that this code block is not the most confusing bit of code in `config.yml` anyway 😛).

##
Just to be clear, in CircleCI each step runs in a new shell, so we can't just source an env file once (as we do on Travis). We have to either source the script on every step (i.e. every `run: ...` entry) or put the vars we want to export into `$BASH_ENV` (which is automatically sourced at the beginning of each step).

IgorMinar replied:
the main concern is readability or maintainability of this code. if you can't think of a better way to write this then let's keep it as is, but if modifying/understanding this block becomes a problem we should try to redo it in a better way.

gkalpak replied:
I tried moving them to a shell script in #26596 (commit: 1857b31c7), but I don't like the end result :grin:

petebacondarwin commented on .circleci/config.yml:
```diff
@@ -41,14 +54,15 @@ anchor_1: &job_defaults
 # See https://discuss.circleci.com/t/1662
 anchor_2: &post_checkout
-  post: git pull --ff-only origin "refs/pull/${CIRCLE_PULL_REQUEST//*pull\//}/merge"
+  post: git pull --ff-only origin "refs/pull/${CI_PULL_REQUEST//*pull\//}/merge"
```
There seems to be a mix of `anchor...` and `var...` definitions. Should we agree on a single convention? Or am I missing something?

gkalpak replied:
I don't think there is a difference as far as yaml syntax is cincerned, but this `config.yml` file seems to use the following convention:
- Use `var_X` for single nodes, that can thus be used as values in sequences. E.g.:
    ```yaml
    listOfThings:
      - *myvar
      - anotherThing
    ```
- Use `anchor_Y` for multiple nodes (e.g. a collection of mappings), that can thus be used with `<<`. E.g.:
    ```yaml
    mapOfThings:
      <<: *myDefaultKeyValuePairs
      anotherKey: anotherValue
      yetAnotherKey: yetAnotherValue
    ```

petebacondarwin replied:
Oh I see.

petebacondarwin commented on .circleci/config.yml:
```diff
@@ -140,38 +154,141 @@ jobs:
+        # Run unit tests for aio/aio-builds-setup
+      - run: ./aio/aio-builds-setup/scripts/test.sh
+        # Deploy angular.io to production (if necessary)
+      - run: echo "export CI_STABLE_BRANCH=$(npm info @angular/core dist-tags.latest | sed -r 's/^\s*([0-9]+\.[0-9]+)\.[0-9]+.*$/\1.x/')" | tee -a $BASH_ENV
```
Shouldn't we move this to the `deploy-production` script?

gkalpak replied:
That's what I wanted to do, but then I couldn't figure out a good way to adapt `deploy-to-firebase.test.sh` to test for different `CI_STABLE_BRANCH` values.
Suggestions welcome :smiley:

petebacondarwin replied:
Hmm, OK. One could have an optional argument for `BRANCH` to `deploy-production` and if not provided it gets it from npm? But I accept that is a bit of a hack.

gkalpak replied:
Yeah, I was tempted to do this, but I was worried changing the behavior of `deploy-to-firebase.sh` like that.
Could definitely be convinced otherwise :Grin:

IgorMinar replied:
move this to the env section / shell script?

gkalpak replied:
I didn't do this to avoid adding the small extra overhead (of getting to npm and fetching the data) to every single step in the config.

IgorMinar replied:
every step? I think it runs once per job. no?

if we are not going to redo the env config then it can stay here.

gkalpak replied:
Yes, every step is a new shell. You can use the `environment` key to set variables for a whole job, but these don't support interpolation (so not very useful).

But if you want to use interpolation, the [recommended way][1] is to echo stuff into `$BASH_ENV`.

[1]: https://circleci.com/docs/2.0/env-vars/#using-bash_env-to-set-environment-variables

gkalpak replied:
Actually my previous comment is wrong :grin: (OK, not wrong, but irrelevant to the issue discussed 😇)
The request to npm would be run once per job (so not a huge overhead indeed). Happy to move this to the `Define environment variables` section at the top, if you think it is better.

petebacondarwin commented on .circleci/config.yml:
```diff
@@ -140,38 +154,141 @@ jobs:
+        # Run unit tests for Firebase redirects
+      - run: yarn --cwd aio redirects-test
+        # Run unit tests for aio/aio-builds-setup
+      - run: ./aio/aio-builds-setup/scripts/test.sh
```
I kind of feel that this should be in the `test_aio_tools` job?

gkalpak replied:
SGTM (I just kept the jobs as they were on Travis) :grin:
I'lll move it over.

petebacondarwin commented on .circleci/config.yml:
```diff
@@ -140,38 +154,141 @@ jobs:
       - run:
           name: Wait for preview and run tests
-          command: |
-            source "./scripts/ci/env.sh" print
```
So we used to have a shell script... for setting the environment vars

gkalpak replied:
🤔

petebacondarwin reviewed (approved):
Looks good!

mary-poppins commented:
You can preview 79345d9 at https://pr26377-79345d9.ngbuilds.io/.

mary-poppins commented:
You can preview 80cc575 at https://pr26377-80cc575.ngbuilds.io/.

mary-poppins commented:
You can preview 66d6c00 at https://pr26377-66d6c00.ngbuilds.io/.

mary-poppins commented:
You can preview 6d7e37d at https://pr26377-6d7e37d.ngbuilds.io/.

IgorMinar commented on .circleci/config.yml.orig:
```diff
@@ -0,0 +1,447 @@
+# Configuration file for https://circleci.com/gh/angular/angular
```
delete?

IgorMinar replied:
remove?

IgorMinar replied:
remove?

IgorMinar commented on .travis.yml:
```diff
@@ -47,16 +39,10 @@ env:
-    - CI_MODE=aio
-    - CI_MODE=aio_local
-    - CI_MODE=aio_e2e AIO_SHARD=0
-    - CI_MODE=aio_e2e AIO_SHARD=1
```
wooooot! 👍

IgorMinar reviewed (approved):
the reset looks good to me. 🤞

IgorMinar reviewed (approved):
there rest looks good to me. 🤞 

thanks for working through this!

IgorMinar commented:
haha.. I see that all of my attempts to submit code reviews eventually made it to github in spite of their outage.. sorry about the duplicates.

mary-poppins commented:
You can preview f262582 at https://pr26377-f262582.ngbuilds.io/.

IgorMinar commented:
I've setup the tokens but had to rename them slightly due to CircleCI validation restrictions. They are now:
- AIO_DEPLOY_TOKEN
- ANGULAR_PAYLOAD_TOKEN

gkalpak committed:
fixup! build: use CI-provider independent variable names

gkalpak committed:
fixup! ci: fix `payload-size.sh`

gkalpak committed:
fixup! ci(docs-infra): move deployment to CircleCI

mary-poppins commented:
You can preview 5a11648 at https://pr26377-5a11648.ngbuilds.io/.

mary-poppins commented:
You can preview 5e58458 at https://pr26377-5e58458.ngbuilds.io/.

gkalpak committed:
fixup! ci(docs-infra): move deployment to CircleCI

gkalpak commented:
Rebased on master, squahsed fixup commits, and added a couple new fixup commits related to using the Firebase tokens.
So, only the current fixup commits have not been reviewed.

gkalpak commented:
Rebased on master, squashed old fixup commits, and added some new fixup commits (mainly related to using the Firebase tokens).
So, only the current fixup commits have not been reviewed.

I suggest we merge this asap and work on possible improvements in #26596.

mary-poppins commented:
You can preview 5b3f708 at https://pr26377-5b3f708.ngbuilds.io/.

IgorMinar reviewed (approved):
yay!
